### ISSUE TITLE
[Develop] Search Overhaul

### ISSUE THREAD

**Issue by nathanlesage** [Zettlr/Zettlr#4231]
Now that 3.0 is (mostly) coming together, and I personally think most areas are already relatively good so far, one area that hasn't received _any_ love in the past three or four years is the search. It's still slow and bad, and thus can actually deserve an improvement – especially since now the amount of notes for most people (me included) begins to approach a volume where the search becomes unbearably slow.

Please use this issue to chime in if you have specific improvements you would like to see. What I'll do in any case is:

- [ ] Improve the speed of the search, allowing for much faster results
- [ ] Improve the syntax used for initiating a search (here I'll be glad for any recommendations of what "makes sense" in the application of Zettlr; currently it's more a very formal library-style search)
- [ ] Allow searching for multi-word tags (maybe with `#"one two"`, but I'm open to syntax suggestions, see above)

**Comment by UFOMelkor** [comment]
Two things I would find useful: Search for broken links and for broken citations. But this might be rather a project linting feature than a search feature.

**Comment by nathanlesage** [comment]
Yup, that's unrelated to search, feel free to open feature requests for that instead,

**Comment by kyaso** [comment]
To significantly boost search speed, one effective solution is to implement a _search index_. I've spent [over a year experimenting](https://github.com/kyaso/Zettlr/compare/v2.2.3+custom.17...v2.2.3+custom.18) with [flexsearch](https://github.com/nextapps-de/flexsearch), and this discussion seems like the perfect chance to share my experience :)

Personally, I rely heavily on Zettelkasten-/Wikilinks in my workflow. For instance, I use "date links" like `[[2023-04-07]]`, and when I click on one, I expect to instantly see all the other notes with the same link (so I can review what I did on that day). With an index-based search, I usually get _instant_ results.

The main drawback of an index-based approach is the increased memory usage. However, nowadays, memory is no longer a scarce resource, and the benefits of a search index outweigh the costs (at least from my experience). We could also consider making this optional in the settings to accommodate different needs.

**Comment by nathanlesage** [comment]
@kyaso Something like this is definitely what I was imagining. Basically, I see there are several construction sites that I'd ideally like to develop and then implement in unison:

1. **The search syntax**. People have complained that, while Zettlr in the meantime detects multi-word tags, you can't search for them, since support for a syntax, e.g., `#"multi words"` or `tag:"multi words"` is missing. Additionally, I feel we should definitely get back to that, since back when I implemented search I was guided by my experience with online library-style searches (those advanced things with operators like NEAR etc) and I don't see this syntax being used in any other editor. So the question is how to refactor the syntax available. I feel by making it less complex we can also cut down on the complexity during the actual search where Zettlr currently has to account for OR and AND operators that sometimes contradict each other.
2. **The search itself**: How can we search more files (without hitting open file handle limits) faster, etc.?
3. **What should be weighted how during search?** Clearly, headings should always have a larger weight, but how can we accomodate this without either heavy taxes on the CPU performance vs. memory usage?
4. **What kind of index should we use for that?** Personally, I think that we should colocate a search index with the FSAL, since the creation of such an index is clearly something that should happen at the same time as parsing a file (i.e. slow startup after the cache is cleared, but only incremental updates afterward). The format of such an index would then be another important question.
5. **What library to use?** I didn't fully jump into your code, I only noticed that you were using `require` which we can't really use in the current setup (but flexsearch supports ES6 syntax). However, I'm fine using a library, but we'd have to do extensive research regarding this to ensure that the library we use (if we use one) …
    1. … isn't too large in terms of bytes added to the binary size
    2. … doesn't come with any performance penalty
    3. … gives us the right amount of flexibility (e.g., where it stores its index, and basic CRUD operations)
    4. … is compatible with whatever search syntax we come up with.

Especially regarding the search syntax, I'm sure that @sensologica can help us out with his UX knowledge here :)

@kyaso That being said, would you be fine with volunteering to implement what we come up with in this discussion? Then I'd kind of keep myself out of the development aspect of this to not interfere here, and reduce myself to the manager :)

**Comment by kyaso** [comment]
> I didn't fully jump into your code, I only noticed that you were using require which we can't really use in the current setup (but flexsearch supports ES6 syntax).

Yeah, my implementation using flexsearch is probably a bit hacky 😅 (back then, I wanted to have a quick working solution). I threw it in here more to show that using a search index is a viable option.

> That being said, would you be fine with volunteering to implement what we come up with in this discussion? Then I'd kind of keep myself out of the development aspect of this to not interfere here, and reduce myself to the manager :)

Uhh, I really appreciate that you are trusting me with this ☺️ I'm excited to help out here, however, I'm afraid I might not have enough bandwidth to tackle _all_ points from this discussion. Maybe, once we have carved out everything, we can break down this issue into smaller ones for the individual requirements/features, and then either one of us (or anybody else interested) could grab issues based on their preference and capacity?

**Comment by nathanlesage** [comment]
> I'm excited to help out here, however, I'm afraid I might not have enough bandwidth to tackle all points from this discussion. Maybe, once we have carved out everything, we can break down this issue into smaller ones for the individual requirements/features, and then either one of us (or anybody else interested) could grab issues based on their preference and capacity?

Don't worry, all of this is volunteering, so I can't and won't demand too much from you! My intention was rather that I would love to give up some responsibility to someone else, and I feel if you kind of "own" the _entire_ feature rebuild, this might be cooler than giving you just scraps. Of course, you can give away any parts of the work you'd like, and I'll of course also help!

But in any case, I know that we first should discuss what to implement before doing so – that's also something I want to improve going forward: more planning, less wild-west hacking on my side, and I think by making this here a team effort this could actually work very well :)

**Comment by sensologica** [comment]
I don't know much about optimizing for performance but here are some ideas that might be relevant to this discussion:

* Visual Studio Code offers search options like "Match Case" and "Match Whole Word". It might be of interest for us to develop similar features as they are quite handy, but also to have "Match Whole Word" enabled by default to eliminate partial matches from surfacing in the results. Like so, searching for "log" would only find instances of the word "log" but not "b**log**", "techno**log**y", etc. 
* Excluded from the above would be matches that actually begin with the search string. Thus "log" should match "**log**ic", "**log**istics", "**log**in", but not ""b**log**", or "bio**log**y". Do people really search for "bio**log**y" by typing "log"? Seems unlikely. So this rule probably makes sense and could declutter search results. At the very least, the matches that begin with the search string should show up above any other ones in the search results as they are likely to be what the user intended to find.
* One could imagine "Match Case" and "Match Whole Word" being configurable in Settings to either be on or off by default, depending on user preference.
* These rules should probably not apply to numbers, punctuation, and maybe other edge cases I am failing to consider at this time.

From a UX standpoint, I would be interested to hear from users about any pain points and workflow issues so we can address them. Please chime in!

**Comment by benniekiss** [comment]
I believe this can be closed with the recent search provider refactor.
